A fully functioning songbook

ericlp opened the pull request:
#### PLZ ROAST ME <3

After way too long it is finally done. Not that it is capable of much, but the project has been restarted multiple times and this time it was actually finished. Here are some of the features that is has.
- Gamma integration
- View all song
- View specific song
- Edit song(only for sexIT)
- Create songs(only for sexIT)
- Delete songs(only for sexIT)

Quite a meagre list but it still exists. 


**Homescreen**
![Screenshot_20210313_010410](https://user-images.githubusercontent.com/31534278/111011054-8be15480-8398-11eb-9ca6-72880683d26b.png)


**View specific song**
![Screenshot_20210313_010439](https://user-images.githubusercontent.com/31534278/111011075-9d2a6100-8398-11eb-92b2-4e7baf8bf7a9.png)


**View specific song with edit-access**
![Screenshot_20210313_010542](https://user-images.githubusercontent.com/31534278/111011084-a3204200-8398-11eb-88ea-132b24148cef.png)


**Edit a song**
![Screenshot_20210313_010604](https://user-images.githubusercontent.com/31534278/111011093-af0c0400-8398-11eb-9676-76a1fa4d00e0.png)

**Confirm deletion of song**
![Screenshot_20210313_010610](https://user-images.githubusercontent.com/31534278/111011117-bdf2b680-8398-11eb-9208-0c24515f11d7.png)


**Create a song**
![Screenshot_20210313_010713](https://user-images.githubusercontent.com/31534278/111011123-c4812e00-8398-11eb-8d85-501eb6816881.png)


**Solves the following issues:**
 - [ #3  ]
 - [ #4  ]
 - [ #5  ]
 - [ #6  ]
 - [ #7  ]
 - [ #11 ]
 - [ #12 ]
 - [ #13 ]
 - [ #14 ]

Portals commented:
Took the liberty to rename the repository.

Before reviewing, I would like to see the application running in production. This often catches smaller mistakes that a review misses. 

Maybe @rosengrenen or @Swexbe can help you with that?

ericlp commented:
Songbook runs fine in production so all the functionality seems to be okey. But I would appreciate if you critiqued to the visual design and code. :heart:

hjorthjort commented:
I think the song view (picture 2) is a bit too narrow. Break-lines are already inserted in songs, so there is no reason to not let the view box be as wide as aesthetically possible, to make reading easier.

The line spacing in the song text seems a bit too large for my taste.

Beautiful work!

ericlp commented:
>  be as wide as aesthetically possible

I am unsure how I should interpret this. Do you mean propose that the panel could be as wide as the screen, with a margin of course?

hjorthjort commented:
> Do you mean that the panel could be as wide as the screen, with a margin of course?

Yes. It's nice that some of the background (the song cards) are visible, but in the screenshots it's a bit much.

ericlp commented:
@hjorthjort I have widened the cards slightly so they don't feel as narrow anymore.

As for the text and line height. That is entirely up to how the song is formatted. The lyrics are structured using markdown. They should be properly formated as soon as sexIT gets around to fix it.

ericlp commented:
I have now fixed all the bugs I can find and cleaned up the code as much as I can bother so plz let the fire rain down on me and burn me until even all the stock of burn heal in palette town isn't enough. <3

ericlp commented:
@rosengrenen come and reck me bby

ericlp commented:
@Portals Do you have any idea why DIgitTranslations or perhaps the loading of the user seems to load for a second while it loads instantly on mat.chalmers.it?

ericlp commented:
@cthit/digit I am satisfied with the PR now and am ready to merge. <3

ViddeM commented:
"Fully functioning" would include an option to export as a printable songbook / print directly from the page as a song. ;)

ericlp commented:
Next PR, thats a future feature bruh

Portals commented on frontend/src/app/components/header/Header.component.jsx:
```diff
@@ -0,0 +1,78 @@
+            <DigitGammaActions
+                customOptions={{
+                    addSong: text.AddSong,
+                    customSignOut: text.SignOut //This should be unnecessary but the prop SignOut didn't work when I tried
```
still no work?

ericlp replied:
Seems to work now, donno what went wrong earlier

Portals commented on frontend/src/common/components/error-card/ErrorCard.component.jsx:
```diff
@@ -0,0 +1,17 @@
+    max-width: 500px;
+`;
+
+export const ErrorCard = ({ message }) => (
```
A title?

ericlp replied:
nah

Portals commented on frontend/src/common/components/error-card/ErrorCard.component.jsx:
```diff
@@ -0,0 +1,17 @@
+    max-width: 500px;
+`;
+
+export const ErrorCard = ({ message }) => (
```
A title?

ericlp replied:
nah

Portals commented on frontend/src/index.js:
```diff
@@ -0,0 +1,12 @@
+import App from "./app";
+
+ReactDOM.render(
+    <DigitProviders>
```
Add `StrictMode`, can catch issues

Portals commented on frontend/src/setupProxy.js:
```diff
@@ -0,0 +1,11 @@
+const { createProxyMiddleware } = require("http-proxy-middleware");
+
+module.exports = function(app) {
+    app.use(
+        "/api",
+        createProxyMiddleware({
+            target: process.env.HTTP_PROXY,
+            changeOrigin: true
+        })
+    );
+};
```
Why do you need this? Why not use `proxy` in `package.json`?

ericlp replied:
I think I tried it but found no success using that method

Portals replied:
Check `mat`. This is unnecessary complexity atm

ericlp replied:
Unsure why it didn't work last time it worked but `proxy` seems to function now.

Portals commented on frontend/src/use-cases/songs/Songs.context.jsx:
```diff
@@ -0,0 +1,90 @@
+        SongTagContext
+    );
+
+    const loadSongs = () => {
```
This should use `useMemo`

ericlp replied:
useMemo doesn't return a function and I can't wrap useMemo in a function either. Isn't it useCallback that I should use?

Portals replied:
Right yeah useCallback

Portals commented on frontend/src/use-cases/songs/Songs.context.jsx:
```diff
@@ -0,0 +1,90 @@
+        }
+    };
+
+    const loadTags = () => {
```
This should use `useMemo`

ericlp replied:
ditto above

Portals commented on frontend/src/use-cases/songs/Songs.jsx:
```diff
@@ -0,0 +1,40 @@
+
+const Songs = () => (
+    <SongTagProvider
+        initialState={InitialSongTagState}
```
InitialState borde tas hand om contextet imo

Portals commented on frontend/src/use-cases/songs/Songs.context.jsx:
```diff
@@ -0,0 +1,90 @@
+
+export const SongTagContext = createContext({});
+
+export const SongTagProvider = ({ reducer, initialState, children }) => (
```
I like to throw in the `Singleton` in the name to signal that this should only be used once in the app.

ericlp replied:
i think it be fine

Portals commented on frontend/src/use-cases/songs/components/song-form/SongForm.utils.jsx:
```diff
@@ -0,0 +1,106 @@
+    )
+});
+
+export const SongFormFields = ({ text, tags }) => {
```
Why not use `DigitEditData`?

ericlp replied:
DigitEditData was wack. I had several problems which was solved when i switched to UseForm. Don't remember them specifically but i think it works pretty nice as of now.

ericlp replied:
I think it mostly had to do with styling. DigitEditData was stupid hard to get the results I wanted while useForm was flexible and nice.

Portals replied:
Valid reason! I will link this to the rdc storybook as a usage example for `useDigitFormField`.

Portals commented on frontend/src/use-cases/songs/screens/create-song/CreateSong.screen.jsx:
```diff
@@ -0,0 +1,26 @@
+
+    return (
+        <SongFormContainer>
+            <div />
```
why?

ericlp replied:
SongFormContainer has a CSS grid with three columns so that the form can be in the middle and the format intructions to the right. But the first column needed some value so it got an empty div. It may not be a pretty solution but it did the job

Portals commented on frontend/src/use-cases/songs/screens/create-song/components/create-song-form/CreateSongForm.component.jsx:
```diff
@@ -0,0 +1,96 @@
+
+export const CreateSongForm = () => {
+    const { tags, loadSongs, loadTags } = useSongTag();
+    let history = useHistory();
```
```suggestion
    const history = useHistory();
```

Portals commented on frontend/src/use-cases/songs/screens/create-song/components/create-song-form/CreateSongForm.component.jsx:
```diff
@@ -0,0 +1,96 @@
+import { ArrowBack } from "@material-ui/icons";
+
+export const CreateSongForm = () => {
+    const { tags, loadSongs, loadTags } = useSongTag();
```
getting a `loadSongs` from `useSongTag` doesn't really make sense

Portals commented on frontend/src/use-cases/songs/screens/create-song/components/create-song-form/CreateSongForm.component.jsx:
```diff
@@ -0,0 +1,96 @@
+
+    useEffect(() => loadTags(), []);
+
+    const performCreate = values => {
```
This can be moved outside the functional component for increased performance

Portals commented on frontend/src/use-cases/songs/screens/edit-song/EditSong.screen.jsx:
```diff
@@ -0,0 +1,64 @@
+const EditSong = () => {
+    let { song_id } = useParams();
+    const [faultySongId, setFaultySongId] = useState(false);
+    const [hasLoadedSong, setHasLoadedSong] = useState(false);
```
Why not just check if `songToEdit` is null or not?

ericlp commented on frontend/src/app/App.routes.js:
```diff
@@ -0,0 +1,17 @@
+import { removeLastSlash } from "../api/utils/api";
+
+export const BASE_ROUTE = "/";
+const SONGS_BASE_ROUTE = "/songs/";
+export const SONGS_EDIT_ROUTE = SONGS_BASE_ROUTE + "edit/";
+export const SONGS_CREATE_ROUTE = SONGS_BASE_ROUTE + "create/";
+
+export const navHome = history => history.push(removeLastSlash(BASE_ROUTE));
+
+export const navViewSong = (history, song_id) =>
+    history.push(removeLastSlash(BASE_ROUTE + song_id));
+
+export const navEditSong = (history, song_id) =>
+    history.push(removeLastSlash(SONGS_EDIT_ROUTE + song_id));
+
+export const navCreateSong = history =>
+    history.push(removeLastSlash(SONGS_CREATE_ROUTE));
```
Any specific reason why this way is bad?

ericlp replied:
How should I only perform the navigation if a request succeeds using `<a></a>`

ericlp replied:
All times I use these javascript is necessary because they are either something provided to ReactDigitComponents or the navigations is conditional based on the success of a request or something of the like

Portals replied:
Use `Link` from `react-router-dom`

Portals replied:
Screen readers like to know what is clickable. Also you can do a CTRL + Click this way

ViddeM replied:
I'm a bit curious, at least for the create navigation it spontaneously feels to me like it probably shouldn't be a link. 
Partly because it might give an error and thus should not redirect but also because I wouldn't expect to be able to CTRL + click on the save button on such a page. But that might just be me. :)

ViddeM replied:
The other two however should probably be links.

ericlp replied:
The dialog lies outside the root element and is therefore outside the router. So Link can't be used in a DigitDialog as I have to wrap the link in a new Router which destroys the navigation.

ericlp replied:
I have converted to use Link where it isn't conditional on whether a request is successful or it isn't in a  DigitDialog.

Portals commented on frontend/src/use-cases/songs/screens/edit-song/EditSong.screen.jsx:
```diff
@@ -0,0 +1,64 @@
+                setHasLoadedSong(true);
+            })
+            .catch(err => {
+                setFaultySongId(true);
```
Well, there can be other issues, not just a 404.

Portals commented on frontend/src/use-cases/songs/screens/edit-song/components/edit-song-form/EditSongForm.component.jsx:
```diff
@@ -0,0 +1,163 @@
+            .then(() => {
+                queueToast({
+                    text: text.DeleteSongSuccessful
+                });
+                navHome(history);
+                loadSongs();
+            })
+            .catch(() => {
+                queueToast({
+                    text: text.DeleteSongFailed
+                });
+                const e = error.response.data.error;
+                setError({ isError: e.isError, message: text[e.message] });
+                if (text[e.message]) {
+                    setSomethingWrong(true);
+                }
+            });
+    };
+
+    const performUpdate = values => {
+        editSong(song.song_id, values)
+            .then(() => {
+                queueToast({
+                    text: text.EditSongSuccessful
+                });
+                navViewSong(history, song.song_id);
+                loadSongs();
+            })
+
+            .catch(error => {
+                queueToast({
+                    text: text.EditSongFailed
+                });
+                const e = error.response.data.error;
+                setError({ isError: e.isError, message: text[e.message] });
+                if (!text[e.message]) {
+                    setSomethingWrong(true);
+                }
+            });
+    };
```
This can also be moved outside the functional component for increased performance.

Portals commented on frontend/src/use-cases/songs/screens/edit-song/components/edit-song-form/EditSongForm.component.jsx:
```diff
@@ -0,0 +1,163 @@
+                                    icon={ArrowBack}
+                                    onClick={() => history.goBack()}
+                                />
+                                <DigitText.Heading6
```
Maybe just use `<DigitText.Text bold>`, might want to delete this component

Portals commented on frontend/src/use-cases/songs/screens/view-songs/ViewSongs.Screen.jsx:
```diff
@@ -0,0 +1,89 @@
+                    setDialogData({ s: song, t: tags });
+                })
+                .catch(err => {
+                    setFaultySongId(true);
```
might be other issues

Portals commented on frontend/src/use-cases/songs/screens/view-songs/components/song-detail/SongDetail.component.jsx:
```diff
@@ -0,0 +1,68 @@
+                    flexWrap={"wrap"}
+                    justifyContent={"space-between"}
+                >
+                    <div>
+                        <DigitText.Text
+                            bold
+                            text={text.Author + ": " + author}
+                        />
+                        <DigitText.Text text={text.Melody + ": " + melody} />
+                    </div>
```
Why use `div`?

ericlp replied:
Just wanted to group them together. Might as well use a Fragment if that is what you are wondering.

ericlp commented on frontend/public/index.html:
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        <!--
+            This comment is used to make react-digit-components work properly
+        -->
+        <noscript id="jss-insertion-point"></noscript>
```
The jss-insertion-point as well?

ericlp replied:
Aaaa, I missread

Portals commented on frontend/src/use-cases/songs/screens/view-songs/components/song-masonry/masonry/useEventListener.js:
```diff
@@ -0,0 +1,26 @@
+import { useEffect, useRef } from "react";
+
+export function useEventListener(eventNames, handler, element = window) {
```
I guess you're using this to handle the layout of the grid, but can't you just use CSS rules?

ericlp replied:
From what I have gathered you need javascript to some degree to get a Masonry pattern.

ericlp replied:
This component is of the smaller ones i have found while browsing for a solution to create a Masonry pattern.

Portals reviewed (changes requested):
Great job!! 

This is a start, might be able to find other smaller things as well, but overall great fucking job man! 

Only focused on the frontend btw.

ericlp commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+
```
git rm be cool

Portals commented on frontend/src/app/components/header/Header.component.jsx:
```diff
@@ -60,12 +55,8 @@ const Header = ({ loading, signIn }) => {
-                        : "https://gamma.chalmers.it"
-                }
-                backendUrl={backendUrl}
+                frontendUrl={process.env.REACT_APP_GAMMA_FRONTEND_URL}
+                backendUrl={process.env.REACT_APP_GAMMA_BACKEND_URL}
```
This won't work. It's pain as well. You have to do this with build arguments, but it's not worth it.

ericlp replied:
I have fixed build arguments now. It may have taken some time but it's pretty neat

ericlp replied:
Have tested with the dockerfile for dev and prod. Both pass their environment variables forward to react.

ericlp replied:
it is now in production and works as expected

Portals commented on .dockerignore:
```diff
@@ -0,0 +1,17 @@
+
+**/node_modules
+
+**/Dockerfile
+**/dev.Dockerfile
+docker-compose.yml
+prod.docker-compose.yml
+
+**/__pycache__
+**/*.pyc
+**/*.pyo
+**/*.pyd
+**/.Python
+**/env
+
+.gitignore
+.git
```
@Gurr1 Is this really necessary since the docker file doesn't copy over these files in `Dockerfile`.

Portals replied:
The frontend Dockerfile has:
```
COPY ./src ./src
COPY package.json .
COPY ./public ./public
```

The backend has:
```
COPY . . 
```

But this `.dockerignore` is not relative to the frontend and backend folder, so this file doesn't really do anything right?

Portals commented on frontend/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM node:latest as build-step
```
Please specify a specific node version, the build can break in the future.

Portals commented on frontend/package.json:
```diff
@@ -0,0 +1,37 @@
+  "name": "frontend",
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
```
Please update dependencies so they match the ones in RDC. The current one look like this:

```
    "peerDependencies": {
        "axios": "0.21.1",
        "react": "17.0.1",
        "react-dom": "17.0.1",
        "react-router-dom": "5.2.0",
        "styled-components": "5.2.1"
    },
```

Portals commented on frontend/public/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  ],
+  "start_url": "./index.html",
+  "display": "standalone",
+  "theme_color": "#000000",
```
Not really the theme color right?

ericlp replied:
I think i might just remove these. They arent used anywhere

Portals commented on frontend/src/app/components/footer/Footer.component.jsx:
```diff
@@ -0,0 +1,28 @@
+`;
+
+export const Footer = () => (
+    <footer>
+        <DigitLayout.Column justifyContent={"center"}>
```
My opinion:
Create a `styled.footer` and make it a flex column. This lessens the amount of elements and depth.

Portals commented on frontend/src/use-cases/songs/screens/edit-song/EditSong.screen.jsx:
```diff
@@ -0,0 +1,64 @@
+            />
+            {songToEdit !== null && (
+                <SongFormContainer>
+                    <div />
```
Why?

ericlp replied:
I suspect that it can be done with the css-effect Grid-column but I fail to make the css apply to get the desired results.

ericlp replied:
i rewrote the structure of that page and this isn't a problem anymore

Portals commented on frontend/src/use-cases/songs/screens/edit-song/components/edit-song-form/EditSongForm.component.jsx:
```diff
@@ -0,0 +1,174 @@
+});
+
+export const EditSongForm = ({ song, setSomethingWrong }) => {
+    let history = useHistory();
```
```suggestion
    const history = useHistory();
```

Portals commented:
@rosengrenen There's a lot of smaller things left to check. I do not agree with approving this.

ViddeM commented on backend/src/process/TagProcess.py:
```diff
@@ -0,0 +1,45 @@
+
+    short_id_res = validate_short_id(tag_id)
+    if short_id_res.is_error:
+        return get_with_error(400, short_id_res.message)
```
In general, perhaps defined constant variables for statuscodes (e.g. `STATUS_BAD_REQUEST = 400`) to avoid using numbers directly in the code.

ViddeM commented on backend/src/web_handler.py:
```diff
@@ -0,0 +1,120 @@
+
+
+class SongRes(Resource):
+    @db_session
```
Preferably don't use @db_sessions outside of command/query files where possible, only having these files communicating with the db (and doing nothing/little else) gives a clean and easy-to-find interface where all database communication is done.

ViddeM reviewed (changes requested):
(Looked briefly at backend only) overall looked pretty good, although there are a few things that I would recommend be changed, see my comments in the code for specifics.

ViddeM reviewed (approved):
Great wurk. git rebase :)

Hoidi reviewed (approved):
Nice code
